### PR TITLE
GeneralsIO socket.io client

### PR DESCRIPTION
**Pull request by revolko**

### CODE REVIEW

File: generals/envs/generalsio_client.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+from socketio import SimpleClient
+
+from generals.agents.agent import Agent
+
+
+class GeneralsBotError(Exception):
+    """Base generals-bot exception
+    TODO: find a place for exceptions
```
Comment (strakam) [inline generals/envs/generalsio_client.py:8]: this is fine here, if the implementation gets too big I can see moving all the code from `/envs/` to elsewhere and creating `exceptions.py`

File: generals/envs/generalsio_client.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    def __init__(self, agent: Agent, user_id: str):
+        super().__init__()
+        self.connect("https://botws.generals.io")
+        self.user_id = user_id
+        self._queue_id = ""
+
+    @property
+    def queue_id(self):
+        if not self._queue_id:
+            raise GeneralsIOClientError("Join the game before waiting for it to start.")
```
Comment (strakam) [inline generals/envs/generalsio_client.py:46]: what does this error mean?

File: examples/client_example.py
Comment (strakam) [inline examples/client_example.py:1]: love this example

File: generals/envs/generalsio_client.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            event = self.receive()[0]
+            if event == "game_start":
+                break
+
+        self._start_game()
+
+    def _start_game(self) -> None:
+        """
+        Triggered after server starts the game.
+        TODO: spawn a new thread in which Agent will calculate its moves
```
Comment (strakam) [inline generals/envs/generalsio_client.py:92]: how hard is this?

**Review by strakam — COMMENTED** [review]
I like that the code is VERY readable and I understand every single function :100: 

idea: lets create a sibling directory next to `envs/` and `core/`, call it `connections/` and put this stuff there. It also probably seems like we should then divide it to:
- `connections/generalsio_client.py`
- `connections/generalsio_client_errors.py` or something like that.  what do you think

File: generals/envs/generalsio_client.py (hunk 2 above)
Comment (revolko) [reply generals/envs/generalsio_client.py:46]: Basically that agent is waiting for game before joining the game. I realise now that the exception message is a bit unfortunate in a way that it is not exception message but more of a suggestion message. I will reword it to something more suitable and put suggestion at the end.

File: generals/envs/generalsio_client.py (hunk 3 above)
Comment (revolko) [reply generals/envs/generalsio_client.py:92]: The tricky part here will be communication between agent and the client threads. Honestly I have very limited experience with it, but I am sure that it is a common problem and will find a lot of good solutions online.

**Comment by revolko** [comment]
I will write down the steps we discussed in the [issue](https://github.com/strakam/generals-bots/issues/94).

I agree to move the functionality to its own folder,  but I suggest naming it `generalsio_client` or `io_client`, what do you think? Moreover, I recommend creating exceptions.py file in `core` and the new folder. The reason for two files is that the `GeneralsBotError` was meant to be as base exception for this library. All other exceptions should extend that base exception. What do you think?

File: generals/remote/generalsio_client.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+    def wait_for_game(self, force_start: bool = True) -> None:
+        """
+        Wait for the game start.
+        :param force_start: If set to True, the Agent will set `Force Start` flag
+        """
+        if force_start:
+            self.emit("set_force_start", (self.queue_id, True))
+
+        while True:
```
Comment (strakam) [inline generals/remote/generalsio_client.py:82]: question/suggestion: given this implementation, if agent is in queue, the loop makes CPU to run on 100% right?
It would be nice if there was something like `time.sleep(0.1)`, wdyt?

File: examples/client_example.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+from generals.agents.random_agent import RandomAgent
+from generals.remote import GeneralsIOClient
+
+
+if __name__ == "__main__":
+    agent = RandomAgent()
+    with GeneralsIOClient(agent, "user_id9l") as client:
+        # register call will fail when given username is already registered
+        client.register_agent("[Bot]MyEpicUsername")
+        client.join_private_game("queueID")
```
Comment (strakam) [inline examples/client_example.py:10]: suggestion: maybe call this function `join_private_lobby` and the one on line 11 as ``join_game`` or something, wdyt?

File: generals/remote/generalsio_client.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+        self._start_game()
+
+    def _start_game(self) -> None:
+        """
+        Triggered after server starts the game.
+        TODO: spawn a new thread in which Agent will calculate its moves
+        """
+        winner = False
+        while True:
```
Comment (strakam) [inline generals/remote/generalsio_client.py:95]: same question here, maybe it is good to look up best practices (i can look it up)

**Review by strakam — COMMENTED** [review]
I love the progress!

File: generals/remote/generalsio_client.py (hunk 4 above)
Comment (revolko) [reply generals/remote/generalsio_client.py:82]: Not really. The `self.receive()` method is active waiting for a message. As long as there are no new messages, the method blocks the whole thread (basically sleeps, I am not sure how it is implemented, but I suspect it to be similar to sleep method --> block thread execution until interrupt, where interrupt in this case will be a new message). Because of that, the while loop is stuck as well.

Comment (strakam) [reply generals/remote/generalsio_client.py:82]: so is the loop required then?

File: examples/client_example.py (hunk 5 above)
Comment (revolko) [reply examples/client_example.py:10]: I have no preference, either works for me. Will rename it.

File: generals/remote/generalsio_client.py (hunk 4 above)
Comment (revolko) [reply generals/remote/generalsio_client.py:82]: Yes. Without it the method would end after receiving a single message. For example `join_game` without a loop would call `start_game` after any random event, such ass chat message or a new bot connecting. The loop is required to read all events until we receive the one breaking the loop.

Comment (strakam) [reply generals/remote/generalsio_client.py:82]: Okay I understand now, thanks for clarification

**Review by strakam — APPROVED** [review]
🚀🚀🚀

File: generals/remote/generalsio_client.py (hunk 4 above)
Comment (revolko) [reply generals/remote/generalsio_client.py:82]: Currently we are using synchronous socket.io implementation. That means that all calls are blocking --> CPU thread is stuck at given call waiting for the response.

There is a asynchronous implementation of socket.io client as well (it will be a part of later improvements of the client). The idea of async is that you create an async call and then `await` it. It might seems the same as sync (the code execution is stuck on the `await` line until the response from the server is received). However, the difference with async is that `await` signals to CPU that it is waiting for something. Based on that information CPU can temporarily assign this thread to a different process, until the interrupt triggering `await` is received and then puts the original awaiting process back on the thread (I don't think that it necessarily must be the same thread).

So yes, sync is kind of wasteful in a sense that it blocks the CPU thread (and CPU cannot assign it to any different process while the original one is waiting, because CPU does not know that the process is waiting). But during the waiting period, the thread does not compute anything, it is in idle state.

I hope it helps with understanding of the whole parallel computation a bit. Feel free to ask more questions.
